traitlets/traitlets.py

```diff
@@ -883,33 +907,131 @@ def __get__(self, inst, cls=None):
+        if names: self.trait_names = names
+
+    def register(self, inst, names=None):
+        """Associate this event with traits on an instance"""
+        pass
+
+    def event_of(self, obj, names=None):
+        """Get the object's trait names this instance is an event of
+
+        Parameters
+        ----------
+        obj: HasTraits instance or class
+            The instance or class being examined
+        name: list, tuple
+            A list of those trait's names. Only those names which
+            this instance is an event of will be returned. If no
+            names are given, then all the trait names of this event
+            are returned instead.
+
+        Returns
+        -------
+        A ``set`` of trait names
+        """
+        named = parse_notifier_name(self.trait_names)
+        tagged = parse_notifier_tags(obj, self.metadata)
+        if names is not None:
+            return set(names).intersection(named + tagged)
+        else:
+            return set(named + tagged)
+
+    def __eq__(self, other):
+        if isinstance(other, TraitEventHandler):
+            return self is other
+        else:
+            return other == self.func
+
+
+class ObserveHandler(TraitEventHandler):
 
-    def __init__(self, names, type):
-        self.trait_names = names
+    caches_instance_resources = True
+
+    def __init__(self, names, tags, type):
+        super(ObserveHandler, self).__init__(names, tags)
         self.type = type
 
     def instance_init(self, inst):
-        inst.observe(self, self.trait_names, type=self.type)
+        self.register(inst)
+
+    def register(self, inst, names=None):
+        if names is not None:
+            matched = self.event_of(inst, names)
+            if len(matched)!=len(names):
+                diff = set(names).difference(matched)
+                m = "Not handling events for %r"
+                raise TraitError(m % list(diff))
+            else:
+                names = list(matched)
+        else:
+            names = self.event_of(inst)
+        inst.observe(self, names, type=self.type)
 
 
-class ValidateHandler(EventHandler):
+class ValidateHandler(TraitEventHandler):
 
-    def __init__(self, names):
-        self.trait_names = names
+    caches_instance_resources = True
 
     def instance_init(self, inst):
-        inst._register_validator(self, self.trait_names)
+        self.register(inst)
+
+    def register(self, inst, names=None):
+        if names is not None:
+            matched = self.event_of(inst, names)
+            if len(matched)!=names:
+                diff = set(names).difference(matched)
+                m = "Not handling events for %r"
+                raise TraitError(m % diff)
+            else:
+                names = list(matched)
+        else:
+            names = self.event_of(inst)
+        inst._register_validator(self, names)
 
 
-class DefaultHandler(EventHandler):
+class DefaultHandler(TraitEventHandler):
 
-    def __init__(self, name):
```

Comment:
`DefaultHandler` is not marked as private, it is importable from top level packages: `from traitlets import default, DefaultHandler` works, 
and despite being a positional argument  you can pass it as a kwarg as for any Python functions not written in C. 

Sidenote, **positional only** argument in Python **do** exist, they are marked by a slash in the function signature:

```
In [3]: input?
Signature: input(prompt=None, /)
```

This is possible only with C Python api, so we can consider that all arguments name in functions signature in python are part of a function api. 

We can check that:  `DefaultHandler(name='foo')` currently work on traitlets stable, and break on with this PR.

Thus regardless of the **intent** of whether this was public or not it **is** a change of API. and it **does** break backward compatibility.